haystack/nodes/sampler/top_p_sampler.py

```diff
@@ -0,0 +1,161 @@
+
+        # compute the similarity scores for these combinations
+        similarity_scores = self.cross_encoder.predict(query_doc_pairs)
+
+        # Apply softmax normalization to the similarity scores
+        probs = np.exp(similarity_scores) / np.sum(np.exp(similarity_scores))
+
+        # Sort the probabilities and calculate their cumulative sum
+        sorted_probs = np.sort(probs)[::-1]
+        cumulative_probs = np.cumsum(sorted_probs)
+
+        # Find the indices with cumulative probabilities that exceed top_p
+        top_p_indices = np.where(cumulative_probs <= top_p)[0]
+
+        # Map the selected indices back to their original indices
+        original_indices = np.argsort(probs)[::-1][top_p_indices]
+        # and select the top_p responses
+        selected_docs = [documents[i] for i in original_indices]
+
+        # if strict is set to False, low p resulted in no documents being selected, then
+        # return at least one document
+        if not selected_docs and not self.strict:
+            highest_prob_indices = np.argsort(probs)[::-1]
+            selected_docs = [documents[highest_prob_indices[0]]]
+
+        # include prob scores in the results
+        if self.top_score_name:
+            for idx, doc in enumerate(selected_docs):
+                doc.meta[self.top_score_name] = "{:.2f}".format(sorted_probs[idx])
+        return selected_docs
+
+    def predict_batch(
+        self,
+        queries: List[str],
+        documents: Union[List[Document], List[List[Document]]],
+        top_p: Optional[float] = None,
+        batch_size: Optional[int] = None,
+    ) -> Union[List[Document], List[List[Document]]]:
+        """
+         - If you provide a list containing a single query...
+
+            - ... and a single list of Documents, the single list of Documents will be re-ranked based on the
+              supplied query.
+            - ... and a list of lists of Documents, each list of Documents will be re-ranked individually based on the
+              supplied query.
+
+
+        - If you provide a list of multiple queries...
+
+            - ... you need to provide a list of lists of Documents. Each list of Documents will be re-ranked based on
+              its corresponding query.
+        """
+        if top_p is None:
+            top_p = self.top_p
+
+        # TODO: add support for batch_size if possible
+
+        if len(queries) == 1 and isinstance(documents[0], Document):
+            return self.predict(queries[0], documents, top_p)  # type: ignore
+
+        if len(queries) == 1 and isinstance(documents[0], list):
+            return [self.predict(queries[0], docs, top_p) for docs in documents]  # type: ignore
+
+        if len(queries) > 1 and isinstance(documents[0], list):
+            return [self.predict(query, docs, top_p) for query, docs in zip(queries, documents)]  # type: ignore
+
+        raise ValueError("Invalid input. Please check the documentation of this method.")
```

Comment:
The error message should describe the expected input format and the given input format.